### DISCUSSION TITLE
Rhino/Grasshopper/ GHopperGIS

### DISCUSSION THREAD

**Discussion by echeverrimontes** [microsoft/vcpkg#36451] [Q&A]
I've been having trouble with an error running the plugin:
1. Solution exception:EPSG PCS/GCS code 6249 not found in EPSG support files.  Is this a valid EPSG coordinate system?
The node I'm having truble with is the vector extractor:
﻿using System;
using System.Collections.Generic;
using Grasshopper.Kernel;

namespace GHopperGIS
{
    public class VectorExtractor : GH_Component
    {
        public override GH_Exposure Exposure => GH_Exposure.primary;

        /// <summary>
        /// Initializes a new instance of the ShapeFileFieldExtractor class.
        /// </summary>
        public VectorExtractor()
          : base( "Vector Extractor", "Vec Extr",
                  "Extracts the desired field from a Vector file",
                  "GHopperGIS", "2. ExtractData" )
        {
        }

        /// <summary>
        /// Registers all the input parameters for this component.
        /// </summary>
        protected override void RegisterInputParams( GH_InputParamManager pManager )
        {

            pManager.AddTextParameter( "Path to file", "Path", "Path to directory where the vector file resides", GH_ParamAccess.item );
            pManager.AddIntegerParameter( "Field to select", "Field", "Input the field you want to select from. Defaults to 1", GH_ParamAccess.item );
            pManager[1].Optional = true;
            pManager.AddIntegerParameter( "Layer to select", "Layer", "Input the layer you want to select from. Defaults to 0", GH_ParamAccess.item );
            pManager[2].Optional = true;
            pManager.AddGenericParameter( "Spatial filter", "SpaFilt", "Input the OSGeo.OGR.Geometry that you want to use as a spatial filter (meaning " +
                                                                       "geometry outside this filter/geometry is culled away). Hint: input the output from " +
                                                                       "the RhinoToWKT component. If nothing is used as input, there will be no spatial " +
                                                                       "filtering.", GH_ParamAccess.item );
            pManager[3].Optional = true;

        }

        /// <summary>
        /// Registers all the output parameters for this component.
        /// </summary>
        protected override void RegisterOutputParams( GH_OutputParamManager pManager )
        {
            
            pManager.AddIntegerParameter( "Number of Features", "NumFeatures", "Outputs the number of features in the layer, files that don't contain this data return -1", GH_ParamAccess.item );
            pManager.AddTextParameter( "Layers", "Lys", "Outputs the layer names", GH_ParamAccess.list );
            pManager.AddTextParameter( "Fields", "Flds", "Gets the fields in the layer", GH_ParamAccess.list );
            pManager.AddTextParameter( "Output field", "Field", "Outputs the field information", GH_ParamAccess.list );
            pManager.AddTextParameter( "Geometry type", "GeoType", "Outputs the geometry type of the vector file", GH_ParamAccess.item );
            pManager.AddTextParameter( "Projection", "Proj", "Outputs the projection (EPSG code) of the vector file", GH_ParamAccess.item );

        }

        /// <summary>
        /// This is the method that actually does the work.
        /// </summary>
        /// <param name="DA">The DA object is used to retrieve from inputs and store in outputs.</param>
        protected override void SolveInstance( IGH_DataAccess DA )
        {

            string isRunning = RunGDAL.output;

            if( isRunning == "GDAL and OGR are up and running!" )
            {

                string input = "";

                DA.GetData( 0, ref input );

                int fieldSelector = 0;

                DA.GetData( 1, ref fieldSelector );

                int layerNumber = 0;
                DA.GetData( 2, ref layerNumber );

                string debug = "";
                OSGeo.OGR.Driver driver = Utilities.GetExtractorDriver( input, ref debug );

                OSGeo.OGR.Geometry geoSpatialFilter = null;
                DA.GetData( 3, ref geoSpatialFilter );

                if( debug != "File does not exist!" )
                { 

                    if( driver == null )
                    {

                        AddRuntimeMessage( GH_RuntimeMessageLevel.Warning, "Driver is null" );

                    }

                    else
                    {

                        OSGeo.OGR.DataSource ds = driver.Open( input, 0 );

                        if( ds == null )
                        {

                            AddRuntimeMessage( GH_RuntimeMessageLevel.Warning, "DataSource is null" );

                        }

                        else
                        {

                            long numberOfFeatures = 0;
                            int numberOfLayers = 0;
                            int fieldCount = 0;

                            numberOfLayers = ds.GetLayerCount();

                            List<string> layerNames = new List<string>();

                            for( int i = 0; i < numberOfLayers; ++i )
                            {

                                OSGeo.OGR.Layer ly = ds.GetLayerByIndex(i);

                                if( ly != null )
                                {

                                   layerNames.Add( ly.GetName() );
                    
                                }

                                else
                                {

                                    layerNames.Add( "Null item!" );

                                }

                            }

                            int layerCount = ds.GetLayerCount();

                            OSGeo.OGR.Layer layer = ds.GetLayerByIndex( layerNumber % layerCount );

                            if( layer != null )
                            { 

                                string geoType = layer.GetGeomType().ToString();                   

                                var layerDefinition = layer.GetLayerDefn();

                                fieldCount = layerDefinition.GetFieldCount();

                                string[] fields = new string[fieldCount];

                                for( int i = 0; i < fieldCount; ++i )
                                {

                                    fields[i] = layerDefinition.GetFieldDefn(i).GetName();

                                } 

                                numberOfFeatures = unchecked( ( int ) layer.GetFeatureCount(0) );

                                fieldCount = layerDefinition.GetFieldCount();

                                OSGeo.OGR.Feature feature = null;

                                // We don't wanna go where we can't don't we?
                                fieldSelector %= fieldCount;

                                if( geoSpatialFilter != null )
                                {

                                    layer.SetSpatialFilter( geoSpatialFilter );

                                }

                                if( numberOfFeatures != -1 && geoSpatialFilter == null )
                                {

                                    string[] fieldOut = new string[numberOfFeatures];   
                        
                                    for( int i = 0; i < numberOfFeatures; ++i )
                                    {
                            
                                        feature = layer.GetFeature(i);
                            
                                        if( feature != null )

                                        fieldOut[i] = feature.GetFieldAsString( fieldSelector );

                                        else 
                                        {

                                            fieldOut[i] = "This feature is null!";

                                        }

                                    }

                                    DA.SetDataList( 3, fieldOut );

                                }

                                else 
                                {

                                    List<string> fieldOut = new List<string>();

                                    feature = layer.GetNextFeature();

                                    while( feature != null )
                                    {

                                        fieldOut.Add( feature.GetFieldAsString( fieldSelector ) );
                                        feature = layer.GetNextFeature();

                                    }

                                    DA.SetDataList( 3, fieldOut );
                        
                                }

                                // Source projections, try to get it automatically otherwise throw runtime error.
                                OSGeo.OSR.SpatialReference source = null;
                                int identifyEPSG = 0;
                                string projection = "";

                                try
                                {

                                    source = layer.GetSpatialRef();
                                    identifyEPSG = source.AutoIdentifyEPSG();
                                    projection = source.GetAttrValue( "AUTHORITY", 1 );

                                }

                                catch
                                {

                                    projection = "The file's EPSG code couldn't be found automatically.";
                                    AddRuntimeMessage( GH_RuntimeMessageLevel.Warning, projection );

                                }

                                DA.SetData( 0, numberOfFeatures );
                                DA.SetDataList( 1, layerNames );
                                DA.SetDataList( 2, fields );
                                DA.SetData( 4, geoType );
                                DA.SetData( 5, projection );

                            }

                            else
                            {

                                AddRuntimeMessage( GH_RuntimeMessageLevel.Warning, "This is layer is null!" );

                            }

                        }

                    }

                }

                else
                {

                    AddRuntimeMessage( GH_RuntimeMessageLevel.Warning, debug );

                }

            }

            else
            {

                AddRuntimeMessage( GH_RuntimeMessageLevel.Error, "Drop into the canvas the component RunGDAL before attempting to do anything else, now hit recompute." );

            }

        }

        /// <summary>
        /// Provides an Icon for the component.
        /// </summary>
        protected override System.Drawing.Bitmap Icon
        {
            get
            {
                //You can add image files to your project resources and access them like this:
                // return Resources.IconForThisComponent;
                return Properties.Resources.VectorExtract;
            }
        }

        /// <summary>
        /// Gets the unique ID for this component. Do not change this ID after release.
        /// </summary>
        public override Guid ComponentGuid
        {
            get { return new Guid("756e3509-48a9-43a6-b2a3-4266d549c3dc"); }
        }
    }
}
If anyone can give me any advice, thank you

**Comment by StarGate-One** [comment]
Need to file a package build failure in vcpkg - see #30604 How to report a package build failure.
